### PR TITLE
feat: add Card component

### PR DESCRIPTION
**Pull request by revolunet**
add Card component

### CODE REVIEW

File: src/Card.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+        assert<Equals<keyof typeof rest, never>>();
+
+        const { Link } = getLink();
+
+        return (
+            <div
+                className={cx(
+                    fr.cx("fr-card"),
+                    enlargeLink === true && fr.cx("fr-enlarge-link"),
```
Comment (garronej) [inline src/Card.tsx:103]: When it's a boolean we don't need to `x === true`, `bool` will do it.  
It's when x isn't a boolean that I prefer explicitely checking for the `undefined`

File: src/Card.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                    iconId !== undefined && fr.cx(iconId),
+                    classes.root,
+                    className
+                )}
+                ref={ref}
+            >
+                <div className={cx(fr.cx("fr-card__body"), classes.body)}>
+                    <div className={cx(fr.cx("fr-card__content"), classes.content)}>
+                        <h3 className={cx(fr.cx("fr-card__title"), classes.title)}>
+                            <Link {...linkProps} className={cx(classes.link)}>
```
Comment (garronej) [inline src/Card.tsx:120]:
```suggestion
                            <Link {...linkProps} className={cx(linkProps.classNames, classes.link)}>
```

File: src/Card.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                            />
+                        </div>
+                        {(badges !== undefined && badges.length && (
+                            <ul className={cx(fr.cx("fr-badges-group"), classes.badges)}>
+                                {badges.map((badge, i) => (
+                                    <li key={i}>{badge}</li>
+                                ))}
+                            </ul>
+                        )) ||
+                            null}
```
Comment (garronej) [inline src/Card.tsx:165]:
```suggestion
                        ))}
```
Ther's no need for a fallback to `null`, `false` is a falid `ReactNode` that is equivalent to `null`.

File: src/Card.tsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                    </div>
+                )) ||
+                    null}
```
Comment (garronej) [inline src/Card.tsx:168]:
```suggestion
                ))}
```

File: stories/Card.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        imageUrl: {
+            "description": "Use any image URL, or none"
+        }
+    },
+    "disabledProps": ["lang"]
+});
+
+export default meta;
+
+const defaultProps = {
```
Comment (garronej) [inline stories/Card.stories.tsx:50]: Nice, I like this factorization 👍

File: stories/getStory.tsx [hunk 6]
```diff
@@ -92,14 +92,17 @@ export function getStoryFactory<Props extends Record<string, any>>(params: {
 
     let isFirstStory = true;
 
-    function getStory(props: Props, params?: { description?: string }): typeof Template {
-        const { description } = params ?? {};
+    function getStory(
+        props: Props,
+        params?: { containerWidth?: number; description?: string }
+    ): typeof Template {
+        const { containerWidth, description } = params ?? {};
```
Comment (garronej) [inline stories/getStory.tsx:99]:
```suggestion
    function getStory(
        props: Props,
        params?: { defaultContainerWidth?: number; description?: string }
    ): typeof Template {
        const { defaultContainerWidth: defaultContainerWidthStoryLevel, description } = params ?? {};
```

File: stories/getStory.tsx [hunk 7]
```diff
@@ -92,14 +92,17 @@ export function getStoryFactory<Props extends Record<string, any>>(params: {
...
 
         const out = Template.bind({});
 
         out.args = {
             "darkMode": window.matchMedia("(prefers-color-scheme: dark)").matches,
-            "containerWidth": defaultContainerWidth ?? 0,
+            "containerWidth": containerWidth ?? defaultContainerWidth ?? 0,
```
Comment (garronej) [inline stories/getStory.tsx:105]:
```suggestion
            "containerWidth": defaultContainerWidthStoryLevel ?? defaultContainerWidth ?? 0,
```

File: src/Card.tsx [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    >;
+} & (CardProps.Default | CardProps.Horizontal);
+
+export namespace CardProps {
+    export type Default = {};
+
+    export type Horizontal = {
+        horizontal: true;
+    };
+}
```
Comment (garronej) [inline src/Card.tsx:65]:
```suggestion
export type CardProps = {
    className?: string;
    title: ReactNode;
    linkProps: RegisteredLinkProps;
    desc?: ReactNode;
    imageUrl?: string;
    imageAlt?: string;
    start?: ReactNode;
    detail?: ReactNode;
    end?: ReactNode;
    endDetail?: ReactNode;
    badges?: ReactNode[]; // todo: restrict to badge component ? these badges are display on the image
    /** where actions can be placed */
    footer?: ReactNode; 
    /** only affect the text */
    size?: "small" | "medium" | "large"; 
    /** make the whole card clickable */
    enlargeLink?: boolean; 
    /** only needed when enlargeMode=true */
    iconId?: FrIconClassName | RiIconClassName; 
    shadow?: boolean;
    background?: boolean;
    border?: boolean;
    grey?: boolean;
    classes?: Partial<
        Record<
            | "root"
            | "title"
            | "card"
            | "link"
            | "body"
            | "content"
            | "desc"
            | "header"
            | "img"
            | "imgTag"
            | "start"
            | "detail"
            | "end"
            | "endDetail"
            | "badges"
            | "footer",
            string
        >
    >;
} & (CardProps.Default | CardProps.Horizontal);

export namespace CardProps {
    export type Default = {
      horizontal?: never;
    };

    export type Horizontal = {
        horizontal: true;
    };
}
```

File: src/Card.tsx [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    start?: ReactNode;
+    detail?: ReactNode;
+    end?: ReactNode;
+    endDetail?: ReactNode;
+    badges?: ReactNode[]; // todo: restrict to badge component ? these badges are display on the image
+    footer?: ReactNode; // where actions can be placed
+    horizontal?: boolean;
+    size?: "small" | "medium" | "large"; // only affect the text
+    enlargeLink?: boolean; // make the whole card clickable
+    iconId?: FrIconClassName | RiIconClassName; // only needed when enlargeMode=true
```
Comment (garronej) [inline src/Card.tsx:31]: What is `enlargeMode`?  Wouldn't it make sence to express this in the type def?

**Review by garronej — COMMENTED** [review]
Thank you <3

File: src/Card.tsx (hunk 2 above)
Comment (revolunet) [reply src/Card.tsx:120]: `className` au singulier non ?

Comment (garronej) [reply src/Card.tsx:120]: Oui pardon!

File: src/Card.tsx (hunk 9 above)
Comment (revolunet) [reply src/Card.tsx]: enlargeMode makes the whole card clickable; what do you mean by `express this in the type def?` ?

Comment (revolunet) [reply src/Card.tsx]: my bad, its not `enlargeMode` but `enlargeLink`

Comment (garronej) [reply src/Card.tsx]: >  enlargeMode makes the whole card clickable; what do you mean by express this in the type def? ?  

```typescript   

type EnlargedLink = {
    enlargeLink: true;
    iconId?: FrIconClassName | RiIconClassName;
};

type NotEnlargedLink = {
    enlargeLink?: false;
    iconId?: never;
}

type Props = .... & (EnlargedLink | NotEnlargedLink);


```

Comment (revolunet) [reply src/Card.tsx]: Yay makes sense thanks

**Comment by garronej** [comment]
Thank you for the adjustments!

**Pull request merged**